### PR TITLE
Support FP32 computations using INT8 weights in TensorFlow Lite XNNPACK delegate

### PR DESCRIPTION
**Pull request by dev0x13**
This PR introduces FP32 computations using per-tensor quantized INT8 weights in XNNPACK delegate. It is done in the same way XNNPACK handles FP16 models. Per-channel quantization support may be supported later.
This PR addresses the issue that TF Lite models quantized in dynamic range mode are not able to employ XNNPACK for faster inference, while most inference computations are generally performed in FP32 using default TF Lite kernels which are basically stealing XNNPACK's job. 
Such models can be produced using TF Lite converter with manually disabled hybrid operators (btw this is a topic for another PR since hybrid kernels seems to be slower than non-hybrid once, and there is no way to turn them off in Python API) generation with dynamic range quantization mode is used: almost all the ops store INT8 weights and use `Dequantize` ops for them at the runtime.
I was successfully employed this technique to significantly speed up the EfficientNet-B5-like model on Android smartphone (Samsung Galaxy S20+, latency: ~800 ms -> ~250 ms).
Please let me know if you want me to share more details on my experiments (benchmark results, precision checking results etc.)  to justify this PR.

### CODE REVIEW

**Comment by abattery** [comment]
@Maratyszcza could you review this P

**Comment by lu-wang-g** [comment]
@Maratyszcza @multiverse-tf Can you please review this PR? Thanks!

**Comment by multiverse-tf** [comment]
> This PR introduces FP32 computations using per-tensor quantized INT8 weights in XNNPACK delegate. It is done in the same way XNNPACK handles FP16 models. Per-channel quantization support may be supported later.
> This PR addresses the issue that TF Lite models quantized in dynamic range mode are not able to employ XNNPACK for faster inference, while most inference computations are generally performed in FP32 using default TF Lite kernels which are basically stealing XNNPACK's job.
> Such models can be produced using TF Lite converter with manually disabled hybrid operators (btw this is a topic for another PR since hybrid kernels seems to be slower than non-hybrid once, and there is no way to turn them off in Python API) generation with dynamic range quantization mode is used: almost all the ops store INT8 weights and use `Dequantize` ops for them at the runtime.

"no-hybrid" means the float32 one here? Could you file an issue regarding the performance slowness in the hybrid one? We will double check our implementation and see why some of our recently landed hybrid optimizations didn't work. Many thanks!

> I was successfully employed this technique to significantly speed up the EfficientNet-B5-like model on Android smartphone (Samsung Galaxy S20+, latency: ~800 ms -> ~250 ms).

The results look really impressive!

> Please let me know if you want me to share more details on my experiments (benchmark results, precision checking results etc.) to justify this PR.

Many thanks for the contribution! I'll wait for @Maratyszcza comments as there're WIP efforts to support quantization in XNNPACK. At the moment, you could try it out adding "--define=xnn_enable_qs8=true" bazel flag when building the XNNPACK library. Admittedly, this WIP support might not work for the hybrid model.

**Comment by dev0x13** [comment]
@multiverse-tf 

> "no-hybrid" means the float32 one here? Could you file an issue regarding the performance slowness in the hybrid one? We will double check our implementation and see why some of our recently landed hybrid optimizations didn't work. Many thanks!

Yes, by hybrid kernels I mean ones similar to [this one](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L1331) (hybrid convolution). I have observed that this kernel is generally slower than pure float32 computations for dynamic range quantized model obtained with TF 2.3 converter, and its performance barely depends on interpreter num threads. However, model obtained with TF 2.4 converter seems to perform faster since hybrid convolution for per-channel quantization case is faster than the one for per-tensor quantization case mentioned above (I believe that TF 2.4 TFLite converter replaces per-tensor dynamic range quantization with the per-channel one, please correct me if I am wrong). Anyway, I will file an issue with the detailed description of experiments I made.

> Many thanks for the contribution! I'll wait for @Maratyszcza comments as there're WIP efforts to support quantization in XNNPACK. At the moment, you could try it out adding "--define=xnn_enable_qs8=true" bazel flag when building the XNNPACK library. Admittedly, this WIP support might not work for the hybrid model.

This PR introduces something different from the pure INT8 inference using XNNPACK. What `xnn_enable_qs8` option introduces is the INT8 computations for **full integer** quantized models, while this PR introduces FP32 computations for **dynamic range** quantized models. So the former speeds up full integer models and the latter speeds up dynamic range quantized models. Hope my explanation makes sense.
Btw I have tested this feature a couple days ago, and seems like it speeds up my full integer quantized ResNet48-like model inference by the factor of 2 comparing to the default Ruy backend, which is really amazing!

**Comment by multiverse-tf** [comment]
> @multiverse-tf
> 
> > "no-hybrid" means the float32 one here? Could you file an issue regarding the performance slowness in the hybrid one? We will double check our implementation and see why some of our recently landed hybrid optimizations didn't work. Many thanks!
> 
> Yes, by hybrid kernels I mean ones similar to [this one](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L1331) (hybrid convolution). I have observed that this kernel is generally slower than pure float32 computations for dynamic range quantized model obtained with TF 2.3 converter, and its performance barely depends on interpreter num threads. However, model obtained with TF 2.4 converter seems to perform faster since hybrid convolution for per-channel quantization case is faster than the one for per-tensor quantization case mentioned above (I believe that TF 2.4 TFLite converter replaces per-tensor dynamic range quantization with the per-channel one, please correct me if I am wrong). Anyway, I will file an issue with the detailed description of experiments I made.

I'm not familiar with this converter part. @daverim, as you have been working on the dynamic range quantization, could you help with this comment? Thx!

> 
> > Many thanks for the contribution! I'll wait for @Maratyszcza comments as there're WIP efforts to support quantization in XNNPACK. At the moment, you could try it out adding "--define=xnn_enable_qs8=true" bazel flag when building the XNNPACK library. Admittedly, this WIP support might not work for the hybrid model.
> 
> This PR introduces something different from the pure INT8 inference using XNNPACK. What `xnn_enable_qs8` option introduces is the INT8 computations for **full integer** quantized models, while this PR introduces FP32 computations for **dynamic range** quantized models. So the former speeds up full integer models and the latter speeds up dynamic range quantized models. Hope my explanation makes sense.

Acked. As @Maratyszcza is on vacation these days (but will be back next week), sorry for being late about the code review. I'll continue to wait for his comments about hybrid kernels regarding the integration into xnnpack delegate.

> Btw I have tested this feature a couple days ago, and seems like it speeds up my full integer quantized ResNet48-like model inference by the factor of 2 comparing to the default Ruy backend, which is really amazing!

Thx for sharing the result :-)

File: tensorflow/lite/delegates/xnnpack/quantization_util.cc [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include <fp16.h>
+#include "tensorflow/lite/kernels/internal/optimized/optimized_ops.h"
+#include "tensorflow/lite/kernels/internal/types.h"
+
+namespace tflite {
+namespace xnnpack {
+
+void DequantizeFloat16(const uint16_t *packed_fp16_data, float *unpacked_fp32_data,
+                       size_t tensor_elements) {
+  for (size_t i = 0; i < tensor_elements; ++i) {
```
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/quantization_util.cc:30]: `std::transform` would be more idiomatic for C++

**Review by Maratyszcza — CHANGES_REQUESTED** [review]
Weights should be generated in three steps:
1. Generate FP32 values.
2. Calculate optimal `scale` for generated values.
3. Quantize the FP32 weights with the calculated `scale` and zero point of 0.
Otherwise, there's a chance that all quantized weights are 0 and the test isn't actually testing much.

Also, IIUC when weights are quantized to INT8, bias should be left as FP32 (technically could be converted to INT32, but it wouldn't save any bytes). cc @multiverse-tf to confirm.

**Comment by gbaned** [comment]
@multiverse-tf  Can you please assist on above comments from @Maratyszcza. Thanks!

File: tensorflow/lite/delegates/xnnpack/quantization_util.cc (hunk 1 above)
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/quantization_util.cc]: Fixed

**Comment by dev0x13** [comment]
@Maratyszcza 

I rewrote INT8 weights generation part.
No changes for FP32 bias instead of INT8 so far due to pending @multiverse-tf review.

File: tensorflow/lite/delegates/xnnpack/quantization_util.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <cstdint>
+#include <cstddef>
+
+#include "tensorflow/lite/kernels/internal/types.h"
+
+namespace tflite {
+namespace xnnpack {
+
+void DequantizeInt8(const int8_t* packed_s8_data, float* unpacked_fp32_data,
```
Comment (multiverse-tf) [inline tensorflow/lite/delegates/xnnpack/quantization_util.h:27]: Add some comments about these 2 functions? Particularly, comment about memory ownership of the "unpacked*" ones?

File: tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc [hunk 3]
```diff
@@ -3253,29 +3259,47 @@ TfLiteIntArray* Delegate::PrepareOpsToDelegate(TfLiteContext* context) {
...
+          case kTfLiteInt8: {
+            TfLiteAffineQuantization *quant_params =
+                static_cast<TfLiteAffineQuantization *>(input_tensor.quantization.params);
+
+            if (quant_params->scale->size != 1) {
+              TF_LITE_KERNEL_LOG(context,
+                                 "unsupported per channel dequantization for INT8 "
+                                 "tensor %d in XNNPACK delegate",
+                                 t);
+              return nullptr;  // Hard error.
```
Comment (multiverse-tf) [inline tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc:3286]: Missing "TfLiteIntArrayFree(nodes_to_delegate);" here?

**Comment by multiverse-tf** [comment]
> Weights should be generated in three steps:
> 
> 1. Generate FP32 values.
> 2. Calculate optimal `scale` for generated values.
> 3. Quantize the FP32 weights with the calculated `scale` and zero point of 0.
>    Otherwise, there's a chance that all quantized weights are 0 and the test isn't actually testing much.
> 
> Also, IIUC when weights are quantized to INT8, bias should be left as FP32 (technically could be converted to INT32, but it wouldn't save any bytes). cc @multiverse-tf to confirm.

I also think that bias should be left as FP32.

File: tensorflow/lite/delegates/xnnpack/conv_2d_tester.cc [hunk 4]
```diff
@@ -265,12 +313,25 @@ std::vector<char> Conv2DTester::CreateTfLiteModel() const {
...
+      tensors.emplace_back(CreateTensor(
+          builder,
+          builder.CreateVector<int32_t>(filter_shape.data(), filter_shape.size()),
+          /*type=*/TensorType_INT8,
+          /*buffer=*/1, /*name=*/0,
+          CreateQuantizationParameters(
+              builder, /*min=*/0, /*max=*/0,
+              builder.CreateVector<float>({filter_scale}),
+              builder.CreateVector<int64_t>({0})),
+          /*is_variable=*/false, /*sparsity=*/sparsity_param));
```
[the remark this answers is not in the document]
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/conv_2d_tester.cc:308]: This line relates to a higher level `CreateTensor` expression, so the indent is correct from my perspective IIUC.

File: tensorflow/lite/delegates/xnnpack/quantization_util.h (hunk 2 above)
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/quantization_util.h:31]: Added comments

File: tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc (hunk 3 above)
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc:3287]: Right, thank you. Fixed

**Comment by dev0x13** [comment]
@multiverse-tf I changed the bias part in tests. Now bias is always stored in FP32.

**Review by multiverse-tf — APPROVED** [review]

**Comment by dev0x13** [comment]
@Maratyszcza @multiverse-tf 
Could anybody review this MR once again? All the requested changes are made and the previous approve doesn't work since there were failed CI jobs (2 most recent commits fixes those). Thank you in advance!

**Review by multiverse-tf — APPROVED** [review]
Sorry for being late for the review. Thx again for your contribution!

**Comment by dev0x13** [comment]
@multiverse-tf Thank you! I can see that some CI jobs are still failing, however these fails seem to be unrelated to this MR changes. Is this an issue for merging?

**Comment by multiverse-tf** [comment]
> @multiverse-tf Thank you! I can see that some CI jobs are still failing, however these fails seem to be unrelated to this MR changes. Is this an issue for merging?

No worries. I'll handle the merge then.

**Comment by dev0x13** [comment]
@multiverse-tf I feel so bad bothering you again, but it seems like there have been some issues during the merge attempt. What can I do to help with this?

**Comment by multiverse-tf** [comment]
> @multiverse-tf I feel so bad bothering you again, but it seems like there have been some issues during the merge attempt. What can I do to help with this?

No worries. I'm actively handling the merge internally. This is a very large PR, and it has invited more comments internally :-) But I think it should be merged by the end of the week. Anyway, I'll let you know when it's merged. Thx for the penitence!

**Comment by dev0x13** [comment]
@multiverse-tf Thank you very much for the clarification and for working on it!

**Comment by multiverse-tf** [comment]
> @multiverse-tf Thank you very much for the clarification and for working on it!

@dev0x13, I'm sorry that the merge could not be achieved this week as the PR enables hybrid model support by default in XNNPACK delegate which requires much more thorough testing internally with other Google products. So to move it forward, we've decided to guard your change with some pre-defined macros. Hope you could understand this. Many thanks in advance, and your contribution again!

**Comment by multiverse-tf** [comment]
> @multiverse-tf
> 
> > "no-hybrid" means the float32 one here? Could you file an issue regarding the performance slowness in the hybrid one? We will double check our implementation and see why some of our recently landed hybrid optimizations didn't work. Many thanks!
> 
> Yes, by hybrid kernels I mean ones similar to [this one](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L1331) (hybrid convolution). I have observed that this kernel is generally slower than pure float32 computations for dynamic range quantized model obtained with TF 2.3 converter, and its performance barely depends on interpreter num threads. However, model obtained with TF 2.4 converter seems to perform faster since hybrid convolution for per-channel quantization case is faster than the one for per-tensor quantization case mentioned above (I believe that TF 2.4 TFLite converter replaces per-tensor dynamic range quantization with the per-channel one, please correct me if I am wrong). Anyway, I will file an issue with the detailed description of experiments I made.
> 
> > Many thanks for the contribution! I'll wait for @Maratyszcza comments as there're WIP efforts to support quantization in XNNPACK. At the moment, you could try it out adding "--define=xnn_enable_qs8=true" bazel flag when building the XNNPACK library. Admittedly, this WIP support might not work for the hybrid model.
> 
> This PR introduces something different from the pure INT8 inference using XNNPACK. What `xnn_enable_qs8` option introduces is the INT8 computations for **full integer** quantized models, while this PR introduces FP32 computations for **dynamic range** quantized models. So the former speeds up full integer models and the latter speeds up dynamic range quantized models. Hope my explanation makes sense.
> Btw I have tested this feature a couple days ago, and seems like it speeds up my full integer quantized ResNet48-like model inference by the factor of 2 comparing to the default Ruy backend, which is really amazing!

@dev0x13, could you share this ResNet48-like tflite model? And on what platform did you get this ~2X speedup? Thx!

**Comment by dev0x13** [comment]
@multiverse-tf

> I'm sorry that the merge could not be achieved this week as the PR enables hybrid model support by default in XNNPACK delegate which requires much more thorough testing internally with other Google products. So to move it forward, we may guard your change with some pre-defined macros. Thanks for your patience, and hope you could understand this. Many thanks again for your contribution!

It's okay, thank you very much for keeping me posted on this! Am I supposed to add these macro guards and a Bazel flag or it's something your team is planning to handle?

> could you share this ResNet48-like tflite model? And on what platform did you get this ~2X speedup? Thx!

The platform is Samsung Galaxy S20+ (global version, not the US one). Important thing to mention is that we are focused on getting the highest inference speed with no regard for resource consumption. So that we heavily employ multithreaded inference for both Ruy and XNNPACK. Currently we use half of the number of physical cores for all threading settings in TFLite interpreter and delegates in assumption that typical Android device chip set implements big.LITTLE architecture (this assumption is kind of statistically proven to be fair since using all physical cores leads to the lower performance).
I am not quite sure about sharing the model as is since it's proprietary, so I need to discuss the sharing options with out team. Why are you asking btw? Is it an unexpected result for XNNPACK being faster than Ruy? :)

**Comment by multiverse-tf** [comment]
> @multiverse-tf
> 
> > I'm sorry that the merge could not be achieved this week as the PR enables hybrid model support by default in XNNPACK delegate which requires much more thorough testing internally with other Google products. So to move it forward, we may guard your change with some pre-defined macros. Thanks for your patience, and hope you could understand this. Many thanks again for your contribution!
> 
> It's okay, thank you very much for keeping me posted on this! Am I supposed to add these macro guards and a Bazel flag or it's something your team is planning to handle?

I'll handle this internally with the merge. Basically, your PR has been imported internally as a pending change, and I'm working on that copy to fix failures that originally blocked the merge, and address internal comments.

> 
> > could you share this ResNet48-like tflite model? And on what platform did you get this ~2X speedup? Thx!
> 
> The platform is Samsung Galaxy S20+ (global version, not the US one). Important thing to mention is that we are focused on getting the highest inference speed with no regard for resource consumption. So that we heavily employ multithreaded inference for both Ruy and XNNPACK. Currently we use half of the number of physical cores for all threading settings in TFLite interpreter and delegates in assumption that typical Android device chip set implements big.LITTLE architecture (this assumption is kind of statistically proven to be fair since using all physical cores leads to the lower performance).

Thx for sharing the testing environment!

> I am not quite sure about sharing the model as is since it's proprietary, so I need to discuss the sharing options with out team. Why are you asking btw? Is it an unexpected result for XNNPACK being faster than Ruy? :)

Well, in order to be accelerated by the XNNPACK delegate (w/ this PR), it looks to me the model has to have a dequantize op for the int8 weight tensor and its output being the input of computation-intensive op, like CONV2D etc.,. However, unlike the fp16 quantization case where we don't have native fp16 kernels, I think the current TFLite converter and the model optimization toolkit will not produce this additional dequantize op for int8-quantized-weights models.  So, I'm just curious how is your model produced.

**Comment by dev0x13** [comment]
@multiverse-tf

Thank you!

> Well, in order to be accelerated by the XNNPACK delegate (w/ this PR), it looks to me the model has to have a dequantize op for the int8 weight tensor and its output being the input of computation-intensive op, like CONV2D etc.,. However, unlike the fp16 quantization case where we don't have native fp16 kernels, I think the current TFLite converter and the model optimization toolkit will not produce this additional dequantize op for int8-quantized-weights models. So, I'm just curious how is your model produced.

Oh, 2x speedup is for `xnn_enable_qs8` flag, i.e. for full-integer quantized model. For dynamic range quantized models I was able to get 3-4x speed up with this PR.
In order to produce a model with dequantize ops I have built TFLite converter from source with manually disabled hybrid operators [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/quantize_weights.cc#L479). From my perspective, an option to enable/disable hybrid kernels in high level TFlite converter API (including Python) is a nice to have feature, so I am think of creating a separate PR for this capability. This seems to be reasonable because currently `use_hybrid_evaluation` option is present in all the functions parameters, but yet hard-coded to `true` [everywhere](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/quantize_weights.cc#L645) for some reason. Would such kind of PR be accepted?

**Comment by multiverse-tf** [comment]
> @multiverse-tf
> 
> Thank you!
> 
> > Well, in order to be accelerated by the XNNPACK delegate (w/ this PR), it looks to me the model has to have a dequantize op for the int8 weight tensor and its output being the input of computation-intensive op, like CONV2D etc.,. However, unlike the fp16 quantization case where we don't have native fp16 kernels, I think the current TFLite converter and the model optimization toolkit will not produce this additional dequantize op for int8-quantized-weights models. So, I'm just curious how is your model produced.
> 
> Oh, 2x speedup is for `xnn_enable_qs8` flag, i.e. for full-integer quantized model. For dynamic range quantized models I was able to get 3-4x speed up with this PR.
> In order to produce a model with dequantize ops I have built TFLite converter from source with manually disabled hybrid operators [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/quantize_weights.cc#L479). From my perspective, an option to enable/disable hybrid kernels in high level TFlite converter API (including Python) is a nice to have feature, so I am think of creating a separate PR for this capability. This seems to be reasonable because currently `use_hybrid_evaluation` option is present in all the functions parameters, but yet hard-coded to `true` [everywhere](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/quantize_weights.cc#L645) for some reason. Would such kind of PR be accepted?

I'm not sure as I didn't know the reasons behind why the option is hard-coded true everywhere (which I guess is probably because we have native int8 hybrid kernels, and those kernels don't need this extra dequantize op at all). But anyway, welcome a PR and let's discuss about it there.

As you may have noticed, this PR has been merged. Basically, to use this feature, one has to compile the library with macro "ENABLE_TFLITE_XNNPACK_DEQUANTIZED_INT8_WEIGHTS" defined, OR set "enable_int8_weights_unpacking" to true when creating the xnnpack delegate explicitly.

**Comment by AIWintermuteAI** [comment]
@dev0x13 Hello, Georgiy! I'm interested in trying your PR with some of my models on Raspberry Pi 4. I have compiled tensorflow lite interpreter pip wheel from commit https://github.com/tensorflow/tensorflow/commit/e5a7475be711e99646fa77dbf8d014bb9d9b6786 with the following parameters:
```
# Build python interpreter_wrapper.
cd "${BUILD_DIR}"
case "${TENSORFLOW_TARGET}" in
  armhf)
    BAZEL_FLAGS="--config=elinux_armhf
      --copt=-march=armv7-a --copt=-mfpu=neon-vfpv4
      --copt=-O3 --copt=-fno-tree-pre --copt=-fpermissive
      --define tensorflow_mkldnn_contraction_kernel=0
      --define=raspberry_pi_with_neon=true
      --define=tflite_pip_with_flex=true
      --define=tflite_with_xnnpack=true"
    ;;
  aarch64)
    BAZEL_FLAGS="--config=elinux_aarch64
      --define tensorflow_mkldnn_contraction_kernel=0
      --define=tflite_pip_with_flex=true
      --define=tflite_with_xnnpack=true
      --define=enable_int8_weights_unpacking=true
      --copt=-O3"
    ;;
  native)
    BAZEL_FLAGS="--copt=-O3 --copt=-march=native
      --define=tflite_pip_with_flex=true
      --define=tflite_with_xnnpack=true"
    ;;
  *)
    BAZEL_FLAGS="--copt=-O3
      --define=tflite_pip_with_flex=true
      --define=tflite_with_xnnpack=true"
    ;;
esac
```
Specifically I plan to test on 64-bit version of Raspberry Pi OS, so I only added  --define=enable_int8_weights_unpacking=true for aarch64. The compilation was successful and I'll try the produced wheel on Monday. Are there any specific requirements for the model to take advantage of this optimization? Or the following conversion code is sufficient?
```
import tensorflow as tf
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_quant_model = converter.convert()
```

**Comment by dev0x13** [comment]
@AIWintermuteAI Hi! Unfortunately it's not that simple. This PR is useful for the restricted set of converted models. Currently there are two hard requirements for them:
1. Model should not contain hybrid ops.
2. Model should only employ per-tensor quantization.

The first requirement currently can be met only by manual disabling hybrid ops in TFLite converter C++ code (e.g. [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/quantize_weights.cc#L479)). Then you should build `tflite_convert` target: `bazel build -c opt tensorflow/lite/python:tflite_convert`.

The second requirement can be met in two ways:
1. Use TensorFlow <2.4 as a base version.
2. Use TensorFlow master as a base version and use [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/python/lite.py#L454) TFLite converter option to force per-tensor quantization instead of per-channel one.
However, I am currently working on the other PR targeting per-channel dynamic quantization support in XNNPACK delegate, so eventually pass away.

**Pull request merged**